### ISSUE TITLE
PLIP: CONTRIBUTING.rst on all repositories

### ISSUE THREAD

**Issue by gforcada** [plone/Products.CMFPlone#1179]
*as I don't know which is the current format to submit a PLIP...*

Problem: a spontaneous (or well seasoned) contributor may not know our acceptance criteria on merging pull requests and generally on contributing to Plone.

Github shows a banner message on top of your pull request submit form:

![contributing](https://cloud.githubusercontent.com/assets/1155680/10543731/4cf8fb22-7428-11e5-9c0d-77486299f745.png)

The idea would be to write down what are our expected best practices that every contribution should follow, add that on ``CONTRIBUTING.rst`` and push that to all Plone repositories.

Best practices:
- README entry, if the change is significant then a paragraph (at least) on the docs folder
- test passes
-  i18n (all strings marked as translatable)
- ...

Any seconder?

**Comment by hvelarde** [comment]
+1 IMO, the changelog entry is necessary only if the changes have some significance for integrators; having changelog entries whit things like "tests fixed" or "fix typo on documentation" only pollutes the changelog and adds nothing to it.

I would add also, squashing commits to the minimum.

and, BTW, IIRC, that information goes on a CONTRIBUTING.rst file and not on CONTRIBUTORS.rst.

**Comment by gforcada** [comment]
@hvelarde thanks for the typo (fixed the file name).

And yes, fully agree, minimal nitpicking changes do not need a changelog entry.

**Comment by tisto** [comment]
+1

**Comment by fgrcon** [comment]
:+1: 
* prs should be based on a understandable documented issue. In many cases one issue needs n pr's (depending on each other), which should be listed up in the issue document (as an anchor for the whole changeset...).

Or is  there an option to create a pr  targeting more than one repository?

**Comment by thet** [comment]
+1 in general

-1 on @hvelarde suggestion to only add info, which is significant to integrators to CHANGES.rst. IMO, every change should go in there, which does not mean, every commit. e.g. if you add a feature, fix a bug, add a test, do some pep8 cleanup, change the css for some element - all of this should go into the changelog - but if you add some follow-up commits to something which is already mentionend in the changelog for the same development version (so, still not released), you don't have to add another changelog entry.

for me, it's the first file i check when i want to see what has changed. it's not only an information for integrators but also for developers.

Maybe we should group them, as other projects do:
- Features
- Bugfixes
- Cleanups

**Comment by frisi** [comment]
@thet grouping is proposed in #1180

**Comment by mauritsvanrees** [comment]
+1. Also +1 on @thet's comments on what ends up in CHANGES.rst

**Comment by thet** [comment]
@frisi tnx, didn't see it.

**Comment by hvelarde** [comment]
@thet IMO, PEP8 clean ups is just another example of what should not be on a changelog entry

**Comment by jensens** [comment]
so i'am happy if people add changelog entries at all. dont increase the barrier introducing sections. KISS!
I'am also +100 to really have any change as an entry in the packages changelog. From a PR merges perspective this is easy to check and also the first i look at when a PR comes in.

We can think about making a second more consumer friendly summary if there are volunteers to write them.

**Comment by jensens** [comment]
I second the intitial idea of having a CONTRIBUTING.rst in any coredev package. If we keep this PLIP in scope thats what it is about.

**Comment by idgserpro** [comment]
The problem of having a CONTRIBUTING.rst in **all** repositories is that a modification in one of them somewhere in time means you have to update that file in all repositories.

We already have http://docs.plone.org/develop/coredev/docs/index.html. If the motivation of this file is to show a warn to an user that some guidelines should be followed, wouldn't it be the case to add just this link to a CONTRIBUTING.rst?

**Comment by mauritsvanrees** [comment]
I am thinking the same thing: put a single link in CONTRIBUTING.rst to a page on docs.plone.org.  It may need to be a much smaller page there, with links to the index for more info.

**Comment by jensens** [comment]
Just placing a link is indeed a good idea. I also think more comprehensive page with 5-7 bullet-points and links to further information is better than pointing directly to the heavy-information page @idgserpro linked above.

**Comment by mauritsvanrees** [comment]
We can discuss what would need to be in the document in https://github.com/plone/buildout.coredev/pull/135

Once merged, the page I created there should end up on http://docs.plone.org/develop/coredev/docs/guidelines.html
Once that is done, we can point to it In CONTRIBUTING.rst.

**Comment by mauritsvanrees** [comment]
I have merged the document that we would link to. I don't know how long it takes before that becomes visible on docs.plone.org, whether that is something automatic or manual.

**Comment by mauritsvanrees** [comment]
The document is live. I have added Products.CMFPlone/CONTRIBUTING.rst and that is now visible when creating a new issue:
https://github.com/plone/Products.CMFPlone/issues/new

**Comment by gforcada** [comment]
@mauritsvanrees thanks for all the work!

**Comment by idgserpro** [comment]
@mauritsvanrees Don't forget to add this change in https://github.com/plone/bobtemplates.plone/ :)

Thanks!

**Comment by gforcada** [comment]
@idgserpro bobtemplates.plone is meant for add-ons not core development, although a generic CONTRIBUTORS.rst would be a good idea nonetheless

**Comment by idgserpro** [comment]
@gforcada you're right. @mauritsvanrees please ignore the request.

Each add-on may have it's own guidelines and don't necessarily needs a contributor's agreement like Plone core.

**Comment by mauritsvanrees** [comment]
I have added `CONTRIBUTING.rst` to all repos in the plone organisation. Or at least in all plone repos that are in the sources.cfg of coredev 5.0.

Perhaps the script I used for this is interesting for others. Or for me when I do a similar thing again. :-) This is run from the src directory:

```
#!/bin/sh
TEXT="Please see http://docs.plone.org/develop/coredev/docs/guidelines.html"
FILENAME="CONTRIBUTING.rst"
COMMIT_MESSAGE="Added CONTRIBUTING.rst.

[ci skip]"

for package in */; do
    echo
    echo "============================================"
    echo
    cd $package
    if test $(git remote -v | grep -c git@github.com:plone) -eq 0; then
        echo "NOPE Skipping $package"
        cd ..
        continue
    fi
    echo "PLONE $package"
    if test -e $FILENAME; then
        cat $FILENAME
        echo "$FILENAME already exists."
        cd ..
        continue
    fi
    git status
    echo
    $(which echo) -n "Create $FILENAME, add, commit and push? [y/n] "
    read -n1 answer
    echo
    echo
    if test "x$answer" != "xy"; then
        echo "NOT doing anything."
        cd ..
        continue
    fi
    echo $TEXT > $FILENAME
    git add $FILENAME
    git ci -m "$COMMIT_MESSAGE"
    git push
    echo DONE
    cd ..
done
```
